pkg/cluster/k8s.go

```diff
@@ -35,199 +28,31 @@ func NewK8sProvisioner() clusterTypes.Provisioner {
-	case *cdc.Recommendation:
-		return k.setUpCDCCluster(s)
-	case *tiflash.Recommendation:
-		return k.setUpTiFlashCluster(s)
-	default:
-		panic("unreachable")
-	}
-}
-
-// TearDown tears down the cluster
-func (k *K8sProvisioner) TearDown(ctx context.Context, spec clusterTypes.ClusterSpecs) error {
-	var err error
-	if !fixture.Context.Purge {
-		return nil
-	}
-	switch s := spec.Defs.(type) {
-	case *tidb.Recommendation:
-		err = k.tearDownTiDBCluster(s)
-	case *binlog.Recommendation:
-		err = k.tearDownBinlogCluster(s)
-	case *cdc.Recommendation:
-		err = k.tearDownCDCCluster(s)
-	case *abtest.Recommendation:
-		err = k.tearDownABtestCluster(s)
-	default:
-		return errors.New("unreachable")
-	}
-	return err
-}
-
-// TODO: move the set up process into tidb package and make it a interface
-func (k *K8sProvisioner) setUpTiDBCluster(recommend *tidb.Recommendation) ([]clusterTypes.Node, []clusterTypes.ClientNode, error) {
-	var (
-		nodes       []clusterTypes.Node
-		clientNodes []clusterTypes.ClientNode
-		err         error
-	)
-	if err := k.CreateNamespace(recommend.NS); err != nil {
-		return nil, nil, errors.New("failed to create namespace " + recommend.NS)
-	}
-	if tc, err := k.TiDB.GetTiDBCluster(recommend.NS, recommend.NS); err == nil {
-		recommend.TidbCluster = tc
-	} else {
-		err = k.TiDB.ApplyTiDBCluster(recommend)
-		if err != nil {
-			return nodes, clientNodes, err
-		}
-	}
-	nodes, err = k.TiDB.GetNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-
-	clientNodes, err = k.TiDB.GetClientNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-
-	return nodes, clientNodes, err
-}
-
-func (k *K8sProvisioner) setUpBinlogCluster(recommend *binlog.Recommendation) ([]clusterTypes.Node, []clusterTypes.ClientNode, error) {
-	var (
-		nodes       []clusterTypes.Node
-		clientNodes []clusterTypes.ClientNode
-		err         error
-	)
-	if err := k.CreateNamespace(recommend.NS); err != nil {
-		return nil, nil, errors.New("failed to create namespace " + recommend.NS)
-	}
-	err = k.Binlog.Apply(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-	nodes, err = k.Binlog.GetNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-	clientNodes, err = k.Binlog.GetClientNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-	return nodes, clientNodes, err
-}
-
-func (k *K8sProvisioner) setUpABTestCluster(recommend *abtest.Recommendation) ([]clusterTypes.Node, []clusterTypes.ClientNode, error) {
-	var (
-		nodes       []clusterTypes.Node
-		clientNodes []clusterTypes.ClientNode
-		err         error
-	)
-	if err := k.CreateNamespace(recommend.NS); err != nil {
-		return nil, nil, errors.New("failed to create namespace " + recommend.NS)
-	}
-	err = k.ABTest.Apply(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
-	}
-	nodes, err = k.ABTest.GetNodes(recommend)
-	if err != nil {
-		return nodes, clientNodes, err
+	// TODO: add iochaos to the new interface
```

Comment:
If you like some configs could be pass down to sub cluster easily, maybe ClusterSpecs can help you. eg. pass it to Apply method.